spec.md

```diff
@@ -3149,6 +3167,227 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+// previous write operation on the volume.
+message GetMetadataAllocatedRequest {
+  // This is the identifier of the snapshot.
+  // This field is REQUIRED.
+  string snapshot_id = 1;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed. It is intended
+  // to be used to continue a previously interrupted call.
+  // The plugins may round down this offset to the nearest alignment
+  // boundary based on the BlockMetadataType used.
+  // This field is REQUIRED.
+  uint64 starting_offset = 2;
+
+  // This is an optional parameter, and if non-zero it specifies the
+  // maximum number of tuples to be returned in each
+  // GetMetadataAllocatedResponse message returned by the RPC stream.
+  // The plugin will determine an appropriate value if 0, and is
+  // always free to send less than the requested value.
+  // This field is OPTIONAL.
+  uint32 max_results = 3;
+
+  // Secrets required by plugin to complete the request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 4 [(csi_secret) = true];
+}
+
+// GetMetadataAllocatedResponse messages are returned in gRPC stream.
+// Cumulatively, they provide information on the allocated data
+// ranges in the snapshot.
+message GetMetadataAllocatedResponse {
+  // This specifies the style used in the BlockMetadata sequence.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // If block_metadata_type is FIXED_LENGTH, then the size_bytes field
+  // of each message in the block_metadata list MUST be constant.
+  // This field is REQUIRED.
+  BlockMetadataType block_metadata_type = 1;
+
+  // This returns the capacity of the underlying volume in bytes.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // This field is REQUIRED.
+  uint64 volume_capacity_bytes = 2;
+
+  // This is a list of data range tuples.
+  // If the value of max_results in the GetMetadataAllocatedRequest
+  // message is non-zero, then the number of entries in this list
+  // MUST be less than or equal to that value.
+  // The byte_offset field of each message MUST be strictly increasing.
+  // The SP MUST NOT return a BlockMetadata message with a byte_offset
+  // less than than the value of starting_offset in the
+  // GetMetadataAllocatedRequest message.
+  // BlockMetadata messages MUST NOT overlap for a given snapshot, i.e.
+  // for any two BlockMetadata messages, A and B, if
+  // A.byte_offset < B.byte_offset then
+  // A.byte_offset + A.size_bytes <= B.byte_offset MUST be true.
+  // This field is OPTIONAL.
+  repeated BlockMetadata block_metadata = 3;
+}
+
+```
+
+##### GetMetadataAllocated Errors
+If the plugin is unable to complete the `GetMetadataAllocated` call successfully it must return a non-OK gRPC code in the gRPC status.
+
+The following conditions are well defined:
```

Comment:
if the snapshot is in the middle of a deletion (which could be a long running op, maybe), what error might this return? if NOT_FOUND then please clarify here.

are there any other snapshot states to consider?

Reply:
The NOT_FOUND error seems the most appropriate from the set of [gRPC status codes](https://grpc.github.io/grpc/core/md_doc_statuscodes.html).

I think it would be fine even if this status code were returned mid-stream when the underlying implementation of this API catches up with the asynchronous deletion happening elsewhere in the storage subsystem.  If it doesn't catch up then all is good for this RPC call, though the client would probably encounter errors later if it were to attempt to access the snapshot concerned.  Either way, there is not much that the CO can do if the snapshot object is deleted from the storage subsystem.

Regarding other snapshot states not involving deletion, the gRPC status codes offer subtle distinctions between FAILED_PRECONDITION, UNAVAILABLE and ABORTED.  I'd expect the SP to be able to map into these states, but I don't see the need to call them out explicitly here.